src/org/rascalmpl/library/lang/html/IO.java

```diff
@@ -12,104 +12,230 @@
-            kids.toArray(a);
-            INode node = factory.node(t.toString(), factory.list(a));
-            node = node.asWithKeywordParameters().setParameters(attributes.pop());
-            stack.peek().add(node);
+        IListWriter w = factory.listWriter();
+        for (Node n : elem.childNodes()) {
+            w.append(toConstructorTree(n));
         }
-        
-        @Override
-        public void handleSimpleTag(Tag t, MutableAttributeSet a, int pos) {
-            INode node = factory.node(t.toString());
-            stack.peek().add(node);
+
+        if (cons.getArity() > 0) {
+            return factory.constructor(cons, new IValue[] { w.done() }, kws);
+        }
+        else {
+            return factory.constructor(cons, new IValue[0], kws);
+        }
+    }
+
+    private IValue toDataConstructor(DataNode node) {
+        return factory.constructor(dataConstructor, factory.string(node.getWholeData()));
+    }
+
+    private IValue toTextConstructor(TextNode elem) {
+        return factory.constructor(textConstructor, factory.string(elem.getWholeText()));
+    }
+
+    /**
+     * Produces a HTML string output from an HTMLElement AST.
+     * 
+     * Why go through all the trouble of building a DOM? The only reason is compliance.
+     * Escapes, encodings, etc. all are maintained by these classes from org.jdom.
+     */
+    public IString writeHTMLString(IConstructor cons, IString charset, IConstructor escapeMode, IBool outline, IBool prettyPrint, IInteger indentAmount, IInteger maxPaddingWidth, IConstructor syntax ) {
+        try (StringWriter out = new StringWriter()) {
+            Document doc = createHTMLDocument(cons);
+            doc = doc.outputSettings(createOutputSettings(charset.getValue(), escapeMode.getName(), outline.getValue(), prettyPrint.getValue(), indentAmount.intValue(), maxPaddingWidth.intValue(), syntax.getName()));
+            return factory.string(doc.outerHtml());
+        }
+        catch (IOException e) {
+            throw RuntimeExceptionFactory.io(e.getMessage());
         }
+    }
+
+    /**
+     * Produces a HTML string output from an HTMLElement AST.
+     * 
+     * Why go through all the trouble of building a DOM? The only reason is compliance.
+     * Escapes, encodings, etc. all are maintained by these classes from org.w3c.dom.
+     */
+    public void writeHTMLFile(ISourceLocation file, IConstructor cons, IString charset, IConstructor escapeMode, IBool outline, IBool prettyPrint, IInteger indentAmount, IInteger maxPaddingWidth, IConstructor syntax ) {
+        try (OutputStream out = URIResolverRegistry.getInstance().getOutputStream(file, false)) {
+            Document doc = createHTMLDocument(cons);
+            doc = doc.outputSettings(createOutputSettings(charset.getValue(), escapeMode.getName(), outline.getValue(), prettyPrint.getValue(), indentAmount.intValue(), maxPaddingWidth.intValue(), syntax.getName()));
+            out.write(doc.outerHtml().getBytes("UTF-8"));
```

Comment:
can't we ask the registry for a Writer or something? so that we don't have to do `.getBytes` on the string?